bpf/aya-bpf/src/helpers.rs

```diff
@@ -701,3 +701,128 @@ pub fn bpf_get_current_pid_tgid() -> u64 {
+    ($fmt:literal $(,)? $($arg:expr),* $(,)?) => {{
+        use $crate::helpers::PrintkArg;
+        // Arguments must be placed on the stack to pass the verifier.
+        let fmt: [u8; $fmt.len() + 1] = *concat_bytes!($fmt, b"\0");
+        let data = [$(PrintkArg::from($arg)),*];
+        $crate::helpers::bpf_printk_impl(&fmt, &data)
+    }};
+}
+
+/// Argument ready to be passed to `printk`.
+#[repr(transparent)]
+#[derive(Copy, Clone)]
+pub struct PrintkArg(u64);
+
+impl PrintkArg {
+    /// Manually construct a `printk` argument.
+    #[inline]
+    pub fn from_raw(x: u64) -> Self {
+        Self(x)
+    }
+}
+
+macro_rules! impl_integer_promotion {
+    ($($ty:ty : via $via:ty),* $(,)?) => {$(
+        /// Create `printk` arguments from integer types.
+        impl From<$ty> for PrintkArg {
+            #[inline]
+            fn from(x: $ty) -> PrintkArg {
+                PrintkArg(x as $via as u64)
+            }
+        }
+    )*}
+}
+
+impl_integer_promotion!(
+  char:  via u64,
+  u8:    via u64,
+  u16:   via u64,
+  u32:   via u64,
+  u64:   via u64,
+  usize: via u64,
+  i8:    via i64,
+  i16:   via i64,
+  i32:   via i64,
+  i64:   via i64,
+  isize: via i64,
+);
+
+/// Construct `printk` arguments from constant pointers.
+impl<T> From<*const T> for PrintkArg {
+    #[inline]
+    fn from(x: *const T) -> Self {
+        PrintkArg(x as usize as u64)
+    }
+}
+
+/// Construct `printk` arguments from mutable pointers.
+impl<T> From<*mut T> for PrintkArg {
+    #[inline]
+    fn from(x: *mut T) -> Self {
+        PrintkArg(x as usize as u64)
+    }
+}
+
+/// Internal helper function for the [`printk!`] macro.
+///
+/// The BPF helpers require the length for both the format string as well as
+/// the argument array to be of constant size to pass the verifier. Const
+/// generics are used to represent this requirement in Rust.
+#[inline]
+#[doc(hidden)]
+pub unsafe fn bpf_printk_impl<const FMT_LEN: usize, const NUM_ARGS: usize>(
+    fmt: &[u8; FMT_LEN],
+    args: &[PrintkArg; NUM_ARGS],
+) -> i64 {
+    // This function can't be wrapped in `helpers.rs` because it has variadic
+    // arguments. We also can't turn the definitions in `helpers.rs` into
+    // `const`s because MIRI believes casting arbitrary integers to function
+    // pointers to be an error.
+    let printk: unsafe extern "C" fn(fmt: *const c_char, fmt_size: u32, ...) -> c_long =
+        mem::transmute(6usize);
+
+    let fmt_ptr = fmt.as_ptr() as *const c_char;
+    let fmt_size = fmt.len() as u32;
+
+    match NUM_ARGS {
+        0 => printk(fmt_ptr, fmt_size),
+        1 => printk(fmt_ptr, fmt_size, args[0]),
+        2 => printk(fmt_ptr, fmt_size, args[0], args[1]),
+        3 => printk(fmt_ptr, fmt_size, args[0], args[1], args[2]),
+        4..=12 => {
+            gen::bpf_trace_vprintk(fmt_ptr, fmt_size, args.as_ptr() as _, (NUM_ARGS * 8) as _)
+        }
+        _ => panic!("the kernel limits printk to a maximum of 12 arguments"),
```

Comment:
Note that this case will actually result in the compiler inlining the panic handler here if `NUM_ARGS > 12`. If the suggested panic handler from the cargo-generate template is used (the one calling `core::hint::unreachable_unchecked), the compiler will optimize away all code paths [dominated by](https://en.wikipedia.org/wiki/Dominator_(graph_theory)) this macro invocation. In my test run this resulted in a funny situation where my whole tracepoint was simply omitted from the binary. Not sure how to best error out here: I wish Rust had proper static asserts.

Reply:
Switched to the "let the kernel deal with it" approach in [379c27c](https://github.com/aya-rs/aya/pull/348/commits/379c27c557bd4828484b76b69c19fae56845e2e1) -- I think it's the best for now. This way we also don't need to hard-code the limit here, in case the kernel folks ever change it.

Reply:
Maybe we can use https://doc.rust-lang.org/stable/std/macro.compile_error.html?

Reply:
Don't think so: since Rust doesn't have something like C++'s `if constexpr (...)` and we instead rely on the optimizer to get rid of the inactive match arms, I expect this to just fail unconditionally. Will try though! 

`assert!` can be const now (by writing `const _: () = assert!(...)`), but `const`s currently can't depend on generic arguments because they are effectively globals under the hood. All a bit unfortunate -- I hope this will be improved over time.

Reply:
Yeah: as expected, this makes compilation fail unconditionally.